vortex-file/src/read/layouts/columnar.rs

```diff
@@ -332,49 +332,63 @@ impl LayoutReader for ColumnarLayoutReader {
                 .map(|name| in_progress_metadata[name].clone()) // TODO(Adam): Some columns might not have statistics
                 .collect::<Vec<_>>();
 
-            Ok(Some(MetadataRead::Value(child_arrays)))
+            Ok(Some(PollRead::Value(child_arrays)))
         } else {
-            Ok(Some(MetadataRead::ReadMore(messages)))
+            Ok(Some(PollRead::ReadMore(messages)))
         }
     }
 
-    fn can_prune(&self, begin: usize, end: usize) -> VortexResult<PruningRead> {
+    fn poll_prune(&self, begin: usize, end: usize) -> VortexResult<PollRead<Prune>> {
         let mut in_progress_guard = self
             .in_progress_prunes
             .write()
             .unwrap_or_else(|poison| vortex_panic!("Failed to write to message cache: {poison}"));
         let selection_range = (begin, end);
-        let in_progress_selection = in_progress_guard
+        let column_prunes = in_progress_guard
             .entry(selection_range)
             .or_insert_with(|| vec![None; self.children.len()]);
         let mut messages = Vec::new();
-        for (i, can_prune_child) in in_progress_selection
+
+        // Check all children to see if they can be pruned.
+        for (i, child_prune_state) in column_prunes
             .iter_mut()
             .enumerate()
             .filter(|(_, a)| a.is_none())
         {
-            match self.children[i].can_prune(begin, end)? {
-                PruningRead::ReadMore(message) => messages.extend(message),
-                PruningRead::Value(is_pruned) => *can_prune_child = Some(is_pruned),
+            match self.children[i].poll_prune(begin, end)? {
+                PollRead::ReadMore(message) => messages.extend(message),
+                PollRead::Value(is_pruned) => {
+                    // If we are short-circuiting and one of the child prunes have been
+                    // pruned, we can immediately prune the range.
+                    if is_pruned == Prune::CanPrune && self.shortcircuit_siblings {
+                        // If any of the children prune, we short-circuit the entire pruning operation.
+                        // We clear the in-progress pruning state and immediately prune this entire
+                        // row range.
+                        in_progress_guard.remove(&selection_range).ok_or_else(|| {
+                            vortex_err!("There were no pruning results and no messages")
+                        })?;
+                        return Ok(PollRead::Value(Prune::CanPrune));
+                    } else {
+                        // Update the state.
+                        *child_prune_state = Some(is_pruned)
+                    }
+                }
```

Comment:
this is the relevant block for short-circuiting, patterned after what we do in the poll_read (formerly read_selection)